vita3k/host/src/pup.cpp

```diff
@@ -0,0 +1,473 @@
+// Vita3K emulator project
+// Copyright (C) 2018 Vita3K team
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+#include <crypto/aes.h>
+#include <host/pup_types.h>
+#include <miniz.h>
+#include <util/fs.h>
+#include <util/string_utils.h>
+
+#include <fstream>
+
+std::map<int, std::string> pup_types = {
+    { 0x100, "version.txt" },
+    { 0x101, "license.xml" },
+    { 0x200, "psp2swu.self" },
+    { 0x204, "cui_setupper.self" },
+    { 0x400, "package_scewm.wm" },
+    { 0x401, "package_sceas.as" },
+    { 0x2005, "UpdaterES1.CpUp" },
+    { 0x2006, "UpdaterES2.CpUp" },
+};
+
+const std::string FSTYPE[] = {
+    "unknown0",
+    "os0",
+    "unknown2",
+    "unknown3",
+    "vs0_chmod",
+    "unknown5",
+    "unknown6",
+    "unknown7",
+    "pervasive8",
+    "boot_slb2",
+    "vs0",
+    "devkit_cp",
+    "motionC",
+    "bbmc",
+    "unknownE",
+    "motionF",
+    "touch10",
+    "touch11",
+    "syscon12",
+    "syscon13",
+    "pervasive14",
+    "unknown15",
+    "vs0_tarpatch",
+    "sa0",
+    "pd0",
+    "pervasive19",
+    "unknown1A",
+    "psp_emulist",
+};
+
+void self2elf(std::string infile, std::string outfile, KeyStore &SCE_KEYS, int klictxt) {
+    std::ifstream filein(infile, std::ios::binary);
+    std::ofstream fileout(outfile, std::ios::binary);
+
+    int npdrmtype = 0;
+
+    char sceheaderbuffer[SceHeader::Size];
+    char selfheaderbuffer[SelfHeader::Size];
+    char appinfobuffer[AppInfoHeader::Size];
+    char verinfobuffer[SceVersionInfo::Size];
+    char controlinfobuffer[SceControlInfo::Size];
+
+    filein.read(sceheaderbuffer, SceHeader::Size);
+    filein.read(selfheaderbuffer, SelfHeader::Size);
+
+    SceHeader sce = SceHeader(sceheaderbuffer);
+    SelfHeader self_hdr = SelfHeader(selfheaderbuffer);
+
+    filein.seekg(self_hdr.appinfo_offset);
+    filein.read(appinfobuffer, AppInfoHeader::Size);
+
+    AppInfoHeader appinfo_hdr = AppInfoHeader(appinfobuffer);
```

Comment:
I see that this header variable are not modified. So make it const so compiler can make optimizations

```suggestion
    const AppInfoHeader appinfo_hdr = AppInfoHeader(appinfobuffer);
```

Reply:
Should be apply with other case.